doc/design/scope.md

```diff
@@ -0,0 +1,58 @@
+# Scope
+
+### Define
+
+Scope is a context to manage Variables. It mainly contains a map from Variable name to Variable. Net will get and update variable throw scope.
+
+```cpp
+class Variable;
+using VariablePtr = std::shared_ptr<Variable>;
```

Comment:
I don't think we need this using alias because in a different scenario we might use different smart pointers.  I'd suggest that we write everything in the definition of `vars_` like:

```cpp
using string = std::string; // So we can switch to other implementation of strings later.
std::map<string, unique_ptr<Variable> > vars_;
```